src/utility/threadpool.cpp

```diff
@@ -102,6 +109,7 @@ void threadpool::shutdown()
 {
     ///////////////////////////////////////////////////////////////////////////
     // Critical Section
+    shutdown_ = true; // prevents deadlock due to threads_ that may spawn during shutdown
```

Comment:
Thread safety: this write is not atomic with the reads, making the implementation not thread safe.